subs/pantry/src/Pantry/Types.hs

```diff
@@ -257,19 +285,54 @@ instance Display PackageLocation where
   display (PLImmutable loc) = display loc
   display (PLMutable fp) = fromString $ toFilePath $ resolvedAbsolute fp
 
+-- | Convert `PackageLocation` to its "raw" equivalent
+--
+-- @since 0.1.0.0
+toRawPL :: PackageLocation -> RawPackageLocation
+toRawPL (PLImmutable im) = RPLImmutable (toRawPLI im)
+toRawPL (PLMutable m) = RPLMutable m
+
+-- | Location for remote packages or archives assumed to be immutable.
+-- as user specifies it i.e. not an exact location
+--
+-- @since 0.1.0.0
+data RawPackageLocationImmutable
+  = RPLIHackage !PackageIdentifierRevision !(Maybe TreeKey)
+  | RPLIArchive !RawArchive !RawPackageMetadata
+  | RPLIRepo    !Repo !RawPackageMetadata
+  deriving (Show, Eq, Ord, Data, Generic)
+
+instance NFData RawPackageLocationImmutable
+instance Store RawPackageLocationImmutable
+
+instance Display RawPackageLocationImmutable where
+  display (RPLIHackage pir _tree) = display pir <> " (from Hackage)"
+  display (RPLIArchive archive _pm) =
+    "Archive from " <> display (raLocation archive) <>
+    (if T.null $ raSubdir archive
+       then mempty
+       else " in subdir " <> display (raSubdir archive))
+  display (RPLIRepo repo _pm) =
+    "Repo from " <> display (repoUrl repo) <>
+    ", commit " <> display (repoCommit repo) <>
+    (if T.null $ repoSubdir repo
+       then mempty
+       else " in subdir " <> display (repoSubdir repo))
+
 -- | Location for remote packages or archives assumed to be immutable.
 --
 -- @since 0.1.0.0
 data PackageLocationImmutable
-  = PLIHackage !PackageIdentifierRevision !(Maybe TreeKey)
+  = PLIHackage !PackageIdentifier !BlobKey !TreeKey
```

Comment:
@qrilka Any reason why this is changed to `PackageIdentifier` from `PackageIdentifierRevision`. I see that in the Raw version it is `PackageIdentifierRevision`.

Reply:
for exact locations we don't need `CabalFileInfo`, , `TreeKey` is enough and e.g. `CFILatest` is not an exact revision :)

Reply:
Ah, I see. Thanks.